### PR TITLE
feat: conventional commit check

### PR DESCRIPTION
**Pull request by GMKrieger**
Updated the contributing file and created a github action that enforces the conventional commit format on commit messages

#31

### CODE REVIEW

**Comment by tuler** [comment]
It would be better if the CONTRIBUTING.md file is the same for all cartesi repos, to facilitate consistency.
But there are sections in there that are repo specific, like scopes and absolute links.

File: .github/workflows/commits.yml
Comment (guidanoli) [inline .github/workflows/commits.yml:17]: It would be great if we could find a GitHub actions that lets us keep a nicer list of commit types and scopes.

File: .github/workflows/commits.yml
Comment (guidanoli) [inline .github/workflows/commits.yml:18]: This is only checking the last commit. We could also try to check all commits in a PR.

File: CONTRIBUTING.md
Comment (gligneul) [inline CONTRIBUTING.md:100]: I think we should call this just offchain. This more generic scope can be used for changes across multiple packages, like when updating dependencies.

File: CONTRIBUTING.md
Comment (gligneul) [inline CONTRIBUTING.md:99]: It is not clear that this list is for onchain scopes. Maybe we should split the list in two.

File: CONTRIBUTING.md
Comment (gligneul) [inline CONTRIBUTING.md:89]: Only if it is a crate for a service.

File: CONTRIBUTING.md
Comment (gligneul) [inline CONTRIBUTING.md:158]:
```suggestion
```

File: .github/SCOPES.md
Comment (gligneul) [inline .github/SCOPES.md:1]: What happens if something is not on this list? For instance, a change to the CI, to the github templates, the the commit for this particular issue?

**Comment by tuler** [comment]
The scopes change a lot over time, during project refactoring.
We should minimize the duplication of valid scopes definition.
I suggest to create a `.scopes` text file, under `.github`, and use that in the CI process to validate commits.
The `.md` files could explain the scopes validation process, but refer to `.github/.scopes` file for the actual valid scopes.

File: .github/workflows/commits.yml
Comment (guidanoli) [inline .github/workflows/commits.yml:24]: Move commit types to file too?

Comment (GMKrieger) [reply .github/workflows/commits.yml:24]: This is standard from conventional commit, it won't change or update no matter what we do in the project. I don't think we need to keep it separate, only if we want to keep this command cleaner

Comment (guidanoli) [reply .github/workflows/commits.yml:24]: Ok, fair enough.

File: CONTRIBUTING.md
Comment (marcelstanley) [inline CONTRIBUTING.md:88]:
```suggestion
The scope is optional, although highly recommended.
However, if you choose to add a scope, it **must** be one of the listed on the file.
```
IMO, breaking lines after each sentence reduces the scope of future changes and makes it easier to read diffs, specially on console.

This is a suggestion that, if possible, I would also apply to the whole document when time allows.

**Comment by marcelstanley** [comment]
Please remember to apply the updated commit rules to this PR itself.

File: CONTRIBUTING.md
Comment (GMKrieger) [reply CONTRIBUTING.md:88]: Sure, no problem.

**Comment by marcelstanley** [comment]
> > Please remember to apply the updated commit rules to this PR itself.
> 
> It already does that, you can look at the checks and see it in action : )

Sure, but I was talking about using the imperative mood on a5e2f6e49f548994bb260775d1eb161ccd0ceb7d ;-)

File: .github/workflows/commits.yml
Comment (gligneul) [inline .github/workflows/commits.yml:19]: This warning shows up in the CI:
```
Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```

File: .github/workflows/commits.yml
Comment (gligneul) [inline .github/workflows/commits.yml:33]: I was checking the output of the job, and for some reason, the error message was messed up:
```
fixup! docs: update contributing.md
feat: conventional commit check
docs: update contributing.md
Invalid commit message: fixup! docs: update contributing.md
feat: conventional commit check
docs: update contributing.md
```
https://github.com/cartesi/rollups/actions/runs/4961673199/jobs/8878760891#step:4:20

File: .github/workflows/commits.yml
Comment (gligneul) [inline .github/workflows/commits.yml:21]: I think it would be better to keep using the pre-made step for checking the conventional commit as a whole. Hence, this step would only check the scopes.

File: .github/workflows/commits.yml
Comment (gligneul) [inline .github/workflows/commits.yml:24]: This regex does not include the optional exclamation mark after the commit type. (See my previous comment.)

File: CONTRIBUTING.md
Comment (gligneul) [inline CONTRIBUTING.md:59]: The optional `!` is missing from here as well.

File: CONTRIBUTING.md
Comment (gligneul) [inline CONTRIBUTING.md:65]: This looks cute.

**Review by gligneul — CHANGES_REQUESTED** [review]

File: .github/workflows/commits.yml
Comment (gligneul) [reply .github/workflows/commits.yml:33]: So, debugging a little bit here, I found that `$commit` == `$commits`.

File: .github/workflows/commits.yml
Comment (GMKrieger) [inline .github/workflows/commits.yml:25]: Ok, so I will leave a comment quickly explaining this regex. It works like this:

If there is no parenthesis, it matches the colon and a previous character.
If there are parenthesis, what is inside MUST be one of the scopes.

The check fails if it doesn't match anything.

Now, this is susceptible to being dodged, e.g. if some writes `feat(askdhaskd: Changed the whole code`, however I thought to myself that trying to generate the perfect regex for all edge cases would be more time consuming than simply looking at the commits and correcting as needed.

If everyone thinks that a perfect regex is more preferable, I will continue iterating on it.

Comment (GMKrieger) [reply .github/workflows/commits.yml:25]: But filtering out this commits would need another regex, right? I would need to check each to commit to see if there's a scope (first solution that comes to mind is checking if there are parenthesis), and run to see if the scope inside is valid.

I don't know if this would simplify the logic by adding another layer of verification. Unless there's something I'm missing, which is possible, I'm not an expert on command line scripts, regex and git manipulation hehehe

Comment (gligneul) [reply .github/workflows/commits.yml:25]: Yes, but it would be a simpler regex. Something like `with_scopes=$(echo -n "$commits" | grep -e '(.*)!\?:')`.

You could write this script in Lua; it will probably be much easier to understand and maintain. Here is a draft from chat.

```
function extractScopesFromFile(filename)
  local scopes = {}
  local file = io.open(filename, "r")
  if not file then
    error("Failed to open file: " .. filename)
  end
  for line in file:lines() do
    local scope = line:match("%S+")
    if scope then
      scopes[scope] = true
    end
  end
  file:close()
  return scopes
end

function extractScopeFromCommit(commitTitle)
  local scope = commitTitle:match("%a+%((.-)%)")
  return scope
end

local filename = ".github/.scopes.txt"
local scopes = extractScopesFromFile(filename)

for line in io.lines() do
  local commitScope = extractScopeFromCommit(line)
  if commitScope and not scopes[commitScope] then
    print("Invalid scope:", commitScope)
    print("Commit:", line)
    os.exit(1)
  end
end
```

Comment (gligneul) [reply .github/workflows/commits.yml:25]: I asked chat to convert this script to bash but it is much harder to understand.

Comment (gligneul) [reply .github/workflows/commits.yml:25]: Another version calling `git log` from inside Lua:
```lua
-- Get commit titles using git log command
local handle = io.popen('git log --pretty=format:"%s" $(git merge-base HEAD origin/main)..HEAD')
local commits = handle:read("*a")
handle:close()

-- Check if the scope matches for each commit
for line in commits:gmatch("[^\r\n]+") do
  local commitScope = extractScopeFromCommit(line)
  if commitScope and not scopes[commitScope] then
    print("Invalid scope:", commitScope)
    print("Commit:", line)
    os.exit(1)
  end
end
```

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:19]: This can be simplified

```suggestion
    return commitTitle:match("%a+%((.-)%)")
```

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:10]: Missing indentation

File: .github/workflows/commits.yml
Comment (guidanoli) [inline .github/workflows/commits.yml:20]: We can write it just as "Lua". It's not an acronym.

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:38]: Inconsistent indentation.
This part uses 2 spaces, while the section above uses 4 spaces.

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:3]: By default, `io.open` opens the file in read mode. We could, therefore, omit the `"r"`.

File: CONTRIBUTING.md
Comment (guidanoli) [inline CONTRIBUTING.md:68]: Adding to @gligneul's comment, we should say it here that the `!` is optional too.

Comment (guidanoli) [reply CONTRIBUTING.md:68]: Okay.

File: CONTRIBUTING.md
Comment (marcelstanley) [inline CONTRIBUTING.md:17]:
```suggestion
Please read and follow our [Code of Conduct](./CODE_OF_CONDUCT.md).
```

File: CONTRIBUTING.md
Comment (marcelstanley) [inline CONTRIBUTING.md:91]:
```suggestion
For information on the scopes of this repository, check the [scopes](./scopes.txt) file for the full list of acceptable scopes.
```

File: .github/.check_scopes.lua
Comment (gligneul) [inline .github/.check_scopes.lua:1]: Missing copyright

**Comment by gligneul** [comment]
Feel free to rebase on top of `next` so we can merge it before the 0.9.0 release.

**Comment by tuler** [comment]
Will `fixup!` commits work with this?

**Comment by gligneul** [comment]
@tuler fixup commits already don't work with conventional commits.

@GMKrieger, the commit message needs to be more accurate because we already have a conventional commit check. It should be something like "feat: check conventional commits scopes."

File: .github/workflows/commits.yml
Comment (gligneul) [inline .github/workflows/commits.yml:5]: The conflict resolution didn't work

Comment (gligneul) [reply .github/workflows/commits.yml:5]: Strange that the CI didn't fail.
Anyway, I'm ready to approve once it is fixed.

Comment (gligneul) [reply .github/workflows/commits.yml:5]: It failed but it is not showing on the PR https://github.com/cartesi/rollups/actions/runs/5082931365

Comment (GMKrieger) [reply .github/workflows/commits.yml]: Ok, It was an IDE error that messed me up. Fixed.

**Review by gligneul — APPROVED** [review]
LGTM. We should get approval from @guidanoli as well.

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:25]: We could omit this line, since, according to the [Lua 5.4 documentation](https://www.lua.org/manual/5.4/manual.html#pdf-file:close)...

> [...] files are automatically closed when their handles are garbage collected [...]

I don't think this is too bad, since we are only opening one file.

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:39]: This can be omitted for the same reason mentioned before.

File: .github/.check_scopes.lua
Comment (gligneul) [reply .github/.check_scopes.lua:25]: It is good practice to close the file since we don't know when the garbage collector will run.

File: .github/.check_scopes.lua
Comment (gligneul) [reply .github/.check_scopes.lua:39]: My previous comment also applies :wink:

File: .github/.check_scopes.lua
Comment (gligneul) [reply .github/.check_scopes.lua:38]: This is not a file; it is a command. Let's leave it as is because we want to iterate over the result once the command terminates successfully. Also, we should change the following line to see if the command did so. Something like `assert(handle:close())`.

File: .github/.check_scopes.lua
Comment (gligneul) [inline .github/.check_scopes.lua:39]:
```suggestion
assert(handle:close())
```

**Review by gligneul — CHANGES_REQUESTED** [review]

File: .github/.check_scopes.lua
Comment (guidanoli) [reply .github/.check_scopes.lua:25]: Yes, this is fair.

File: .github/.check_scopes.lua
Comment (guidanoli) [reply .github/.check_scopes.lua:38]: Doesn't `io.popen` return a file handle?
Why can't we iterate over the result while the command is running?
My idea is the following:

```lua
local handle = io.popen('git log ...')
for line in handle:lines() do
  local commitScope = extractScopeFromCommit(line)
  -- (...)
end
```

Comment (gligneul) [reply .github/.check_scopes.lua:38]: This works if the git command succeeds. However, if git returns an error, the code will process error lines as if they were commits.
Here is my idea
```lua
local handle = io.popen('git log --pretty=format:"%s" $(git merge-base HEAD origin/main)..HEAD')
local commits = handle:read("*a")
assert(handle:close())
-- At this point, we know for sure that the git command succeeded
```

Comment (guidanoli) [reply .github/.check_scopes.lua:38]: I see your point now. Thanks for the clarification :-)
In this case, I would suggest removing the `*` in the call `:read("*a")`, as it is ignored by Lua since 5.3.

**Review by gligneul — DISMISSED** [review]

File: .github/.check_scopes.lua
Comment (guidanoli) [inline .github/.check_scopes.lua:21]: Rather than just checking if `scopes[scope] == nil`, we should assert that this is not the case, because it would indicate that something is wrong with the `.scopes.txt` file.

**Comment by GMKrieger** [comment]
> Not to add even more to a 3-weeks PR, but IMHO checking scopes is overrated, for a 3-weeks PR 🫠

It was supposed to be a quick update, but everyone had strong opinions for the contributing document going forward, from styling to function to format, so it took a while for everyone to voice their wishes and the updates be done.

Since this is a point of entry for the project going forward, specially in an open source world, it's understandable that we should be all aligned and happy with this, so that there are fewer stresses and maintenance in the future.

File: .github/.check_scopes.lua
Comment (guidanoli) [reply .github/.check_scopes.lua:21]: Fair enough. Either way, checking if `scopes[scope] == nil` is useless if the purpose is just to collect the set of valid scopes from the scopes file.

**Review by gligneul — APPROVED** [review]

**Pull request merged**
